[UT]lots of netty client ClosedChannelException in UT logs

The exception is thrown when a netty client closes its channel. It sends a downstream event but the server is already down. Here are the details.

```
INFO] [02/03/2015 09:24:49.181] [Client$] closing netty client netty-client-HostPort(192.168.1.74,48598)...
[ERROR] [02/03/2015 09:24:49.182] [Client$] failed to send requests to /192.168.1.74:48598
java.nio.channels.ClosedChannelException
        at org.jboss.netty.channel.socket.nio.AbstractNioWorker.cleanUpWriteBuffer(AbstractNioWorker.java:405)
        at org.jboss.netty.channel.socket.nio.AbstractNioWorker.close(AbstractNioWorker.java:373)
        at org.jboss.netty.channel.socket.nio.NioClientSocketPipelineSink.eventSunk(NioClientSocketPipelineSink.java:58)
        at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendDownstream(DefaultChannelPipeline.java:779)
        at org.jboss.netty.handler.codec.oneone.OneToOneEncoder.handleDownstream(OneToOneEncoder.java:54)
        at org.jboss.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:591)
        at org.jboss.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:582)
        at org.jboss.netty.channel.Channels.close(Channels.java:812)
        at org.jboss.netty.channel.AbstractChannel.close(AbstractChannel.java:197)
        at org.apache.gearpump.transport.netty.Client.org$apache$gearpump$transport$netty$Client$$close(Client.scala:138)
        at org.apache.gearpump.transport.netty.Client.postStop(Client.scala:145)
        at akka.actor.Actor$class.aroundPostStop(Actor.scala:475)
        at org.apache.gearpump.transport.netty.Client.aroundPostStop(Client.scala:36)
        at akka.actor.dungeon.FaultHandling$class.akka$actor$dungeon$FaultHandling$$finishTerminate(FaultHandling.scala:210)
        at akka.actor.dungeon.FaultHandling$class.terminate(FaultHandling.scala:172)
        at akka.actor.ActorCell.terminate(ActorCell.scala:369)
        at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:462)
        at akka.actor.ActorCell.systemInvoke(ActorCell.scala:478)
        at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:263)
        at akka.dispatch.Mailbox.run(Mailbox.scala:219)
        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:393)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
[ERROR] [02/03/2015 09:24:49.185] [Client$] failed to send requests to /192.168.1.74:48598
java.nio.channels.ClosedChannelException
        at org.jboss.netty.channel.socket.nio.AbstractNioWorker.cleanUpWriteBuffer(AbstractNioWorker.java:405)
        at org.jboss.netty.channel.socket.nio.AbstractNioWorker.close(AbstractNioWorker.java:373)
        at org.jboss.netty.channel.socket.nio.NioClientSocketPipelineSink.eventSunk(NioClientSocketPipelineSink.java:58)
        at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendDownstream(DefaultChannelPipeline.java:779)
        at org.jboss.netty.handler.codec.oneone.OneToOneEncoder.handleDownstream(OneToOneEncoder.java:54)
        at org.jboss.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:591)
        at org.jboss.netty.channel.DefaultChannelPipeline.sendDownstream(DefaultChannelPipeline.java:582)
        at org.jboss.netty.channel.Channels.close(Channels.java:812)
        at org.jboss.netty.channel.AbstractChannel.close(AbstractChannel.java:197)
        at org.apache.gearpump.transport.netty.Client.org$apache$gearpump$transport$netty$Client$$close(Client.scala:138)
        at org.apache.gearpump.transport.netty.Client.postStop(Client.scala:145)
        at akka.actor.Actor$class.aroundPostStop(Actor.scala:475)
        at org.apache.gearpump.transport.netty.Client.aroundPostStop(Client.scala:36)
        at akka.actor.dungeon.FaultHandling$class.akka$actor$dungeon$FaultHandling$$finishTerminate(FaultHandling.scala:210)
        at akka.actor.dungeon.FaultHandling$class.terminate(FaultHandling.scala:172)
        at akka.actor.ActorCell.terminate(ActorCell.scala:369)
        at akka.actor.ActorCell.invokeAll$1(ActorCell.scala:462)
        at akka.actor.ActorCell.systemInvoke(ActorCell.scala:478)
        at akka.dispatch.Mailbox.processAllSystemMessages(Mailbox.scala:263)
        at akka.dispatch.Mailbox.run(Mailbox.scala:219)
        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:393)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
```

It seems to be a long standing [netty issue](https://issues.jboss.org/browse/NETTY-412).

--- a/core/src/main/scala/org/apache/gearpump/transport/netty/Client.scala
+++ b/core/src/main/scala/org/apache/gearpump/transport/netty/Client.scala
@@ -19,6 +19,7 @@
 package org.apache.gearpump.transport.netty
 
 import java.net.{ConnectException, InetSocketAddress}
+import java.nio.channels.ClosedChannelException
 import java.util.Random
 import java.util.concurrent.TimeUnit
 
@@ -184,7 +185,9 @@ object Client {
 
     override def exceptionCaught(ctx: ChannelHandlerContext, event: ExceptionEvent) {
       event.getCause match {
-        case ex : ConnectException => Unit
+        case ex: ConnectException => Unit
+        case ex: ClosedChannelException =>
+          LOG.warn("exception found when trying to close netty connection", ex.getMessage)
         case ex => LOG.error("Connection failed " + name, ex)
       }
     }

--- a/streaming/src/test/scala/org/apache/gearpump/streaming/appmaster/AppMasterSpec.scala
+++ b/streaming/src/test/scala/org/apache/gearpump/streaming/appmaster/AppMasterSpec.scala
@@ -30,6 +30,7 @@ import org.apache.gearpump.cluster.appmaster.AppMasterRuntimeEnvironment
 import org.apache.gearpump.cluster.master.{AppMasterRuntimeInfo, MasterProxy}
 import org.apache.gearpump.cluster.scheduler.{Relaxation, Resource, ResourceAllocation, ResourceRequest}
 import org.apache.gearpump.partitioner.HashPartitioner
+import org.apache.gearpump.streaming.ExecutorToAppMaster.RegisterTask
 import org.apache.gearpump.streaming.task._
 import org.apache.gearpump.streaming.{AppDescription, TaskDescription}
 import org.apache.gearpump.util.ActorSystemBooter.RegisterActorSystem
@@ -119,7 +120,7 @@ class AppMasterSpec extends WordSpec with Matchers with BeforeAndAfterEach with 
       watcher.expectTerminated(mockMasterProxy)
 
       mockMasterProxy = getActorSystem.actorOf(Props(new MasterProxy(List(mockMaster.ref.path))), AppMasterSpec.MOCK_MASTER_PROXY)
-      mockMaster.expectMsgType[RegisterAppMaster]
+      mockMaster.expectMsgClass(15 seconds, classOf[RegisterAppMaster])
     }
 
     "launch executor and task properly" in {

